DirectProgramming/DPC++FPGA/Tutorials/Features/ac_types/ac_int/src/ac_int.cpp

```diff
@@ -0,0 +1,350 @@
+    accessor y{inp2, h, read_only};
+    accessor res{result, h, write_only};
+    h.single_task<Mult>([=] { res[0] = x[0] * y[0]; });
+  });
+  q.wait();
+}
+
+void TestShiftLeft(queue &q, const ac_int14 &a, const ac_int14 &b,
+                   ac_int14 &c) {
+  buffer<ac_int14, 1> inp1(&a, 1);
+  buffer<ac_int14, 1> inp2(&b, 1);
+  buffer<ac_int14, 1> result(&c, 1);
+
+  q.submit([&](handler &h) {
+    accessor x{inp1, h, read_only};
+    accessor y{inp2, h, read_only};
+    accessor res{result, h, write_only};
+    h.single_task<ShiftLeft>([=] { res[0] = x[0] << y[0]; });
+  });
+  q.wait();
+}
+
+// Note how the shift amount is specified with a smaller ac_int than in
+// TestShiftLeft above
+void TestEfficientShiftLeft(queue &q, const ac_int14 &a, const ac_uint4 &b,
+                            ac_int14 &c) {
+  buffer<ac_int14, 1> inp1(&a, 1);
+  buffer<ac_uint4, 1> inp2(&b, 1);
+  buffer<ac_int14, 1> result(&c, 1);
+
+  q.submit([&](handler &h) {
+    accessor x{inp1, h, read_only};
+    accessor y{inp2, h, read_only};
+    accessor res{result, h, write_only};
+    h.single_task<EfficientShiftLeft>([=] { res[0] = x[0] << y[0]; });
+  });
+  q.wait();
+}
+
+// The method
+//      x = y.slc<M>(n)
+// is equivalent to the VHDL behavior of
+//      x := y((M+n-1) downto n);
+// Note that only static bit widths are supported
+void TestGetBitSlice(queue &q, const ac_int14 &a, ac_uint4 &b, const int lsb) {
+  buffer<ac_int14, 1> inp1(&a, 1);
+  buffer<int, 1> inp2(&lsb, 1);
+  buffer<ac_uint4, 1> result(&b, 1);
+
+  q.submit([&](handler &h) {
+    accessor x{inp1, h, read_only};
+    accessor y{inp2, h, read_only};
+    accessor res{result, h, write_only};
+    h.single_task<GetBitSlice>([=] { res[0] = x[0].slc<4>(y[0]); });
+  });
+  q.wait();
+}
+
+// There is a set_slc(int lsb, const ac_int<W, S> &slc) which allows the user to
+// set a bit slice as shown in the example below.
+void TestSetBitSlice(queue &q, ac_int14 &a, const ac_int4 &b, int lsb) {
+  buffer<ac_int14, 1> inp1(&a, 1);
+  buffer<ac_int4, 1> inp2(&b, 1);
+  buffer<int, 1> inp3(&lsb, 1);
+
+  q.submit([&](handler &h) {
+    accessor x{inp1, h, read_write};
+    accessor y{inp2, h, read_only};
+    accessor lsb_accessor{inp3, h, read_only};
+
+    h.single_task<SetBitSlice>([=] {
+      // The set_slc method does not need to have a width specified as a
+      // template argument since the width is inferred from the width of the
+      // argument x
+      x[0].set_slc(lsb_accessor[0], y[0]);
+
+      // Bits can also be individually set as follows:
+      x[0][3] = 0;
+      x[0][2] = 0;
+      x[0][1] = 0;
+      x[0][0] = 0;
+    });
+  });
+  q.wait();
+}
+
+int main() {
+#if defined(FPGA_EMULATOR)
+  ext::intel::fpga_emulator_selector selector;
+#else
+  ext::intel::fpga_selector selector;
+#endif
+  bool passed = true;
+
+  try {
+    // create the SYCL device queue
+    queue q(selector, dpc_common::exception_handler);
+
+    // Initialize two random ints
```

Comment:
I would not use rand() at all... I would want to explicitly show a negative value and a positive value.
Also, why not use `ac_int` types for `t1` and `t2` instead of native `int` types?

Reply:
> why not use ac_int types for t1 and t2 instead of native int types?
I think it is to emphasize that the int vs ac_int comparison

How about I replace it with :
```
int t1 = 624
int t2 = -254
{
  int golden = t1  + t2;
  ac_int14 a = t1;
  ac_int14 b = t2;
  ac_int15 c;
  TestAdd(q, a, b, c);
  ...
}
```

Reply:
can this be written as

```c++
ac_int14 a = 624;
ac_int14 b = -254;
{
  int golden = a  + b;
  ac_int15 c;
  TestAdd(q, a, b, c);
  ...
}
```